### DISCUSSION TITLE
Here's what all the fuss is about....

### DISCUSSION THREAD

**Discussion by MickTheMechanic** [ImpulseAdventure/GUIslice#341] [Show and tell]
First of all I want to thank you guys for the help, support and most of all for creating such a cool GUI building tool.

I have opened a few issues here and I thought I'd share what all the fuss has been about. Its not completely finished, but it's not far off. 

I am a hobby racer and I have been looking for a DIY tire pyrometer with memery function, an important tool if you're serious about any kind of fine tuning of the chassis/suspension. The only thing that I have found is[ this project from mouser.](https://eu.mouser.com/applications/automotive-tire-pyrometer-assembly/) It would be fine, but I thought I could do better than that. [Commercial items](http://www.longacreracing.com/products.aspx?itemid=1721&pagetitle=Tire-Pyrometers) (wit memory) cost between 200-500 euros.

The pyrometer is for checking internal temps (inside the tread), but sometimes just a quick look at surface temps can help alot, especially if there isn't enough time to take a full log of all internal temps in the pits. For this reason a thermal camera is ideal. Once again, anything kind of decent will cost a few hundred.

Another must have tool is a tire pressure gauge, good quality digital items are also not cheap.

Here, using your library I have combined all three of these for a fraction of the price, and when everything is finished the project will be made open source. The aim is to try and balance the playing field in grass roots motorsport, give the little guy a chance with tools that are usually limited to those with much bigger budgets.

Here are some photos of how the UI looks:

![IMG_0234](https://user-images.githubusercontent.com/44751666/103695734-6e110100-4f9d-11eb-9f4a-0327efb383fd.jpg)
Screensaver

![IMG_0224](https://user-images.githubusercontent.com/44751666/103695776-7e28e080-4f9d-11eb-9d4b-f26a779df31d.jpg)
Home screen

![IMG_0226](https://user-images.githubusercontent.com/44751666/103695816-900a8380-4f9d-11eb-8909-e47d0b856698.jpg)
Pressure gauge

![IMG_0227](https://user-images.githubusercontent.com/44751666/103695853-9b5daf00-4f9d-11eb-828b-5dba3e19f1b5.jpg)
Temperature gauge

![IMG_0228](https://user-images.githubusercontent.com/44751666/103695872-a6184400-4f9d-11eb-913b-559cbc700b96.jpg)
![IMG_0229](https://user-images.githubusercontent.com/44751666/103695894-af091580-4f9d-11eb-970b-36467d179402.jpg)
Thermal camera

![IMG_0230](https://user-images.githubusercontent.com/44751666/103695914-b9c3aa80-4f9d-11eb-8367-b81e23172ea9.jpg)
![IMG_0231](https://user-images.githubusercontent.com/44751666/103695928-c6480300-4f9d-11eb-8ea9-73ed92b783af.jpg)
![IMG_0233](https://user-images.githubusercontent.com/44751666/103695942-ccd67a80-4f9d-11eb-990d-91d3ff2a611d.jpg)
Data log

Cad model for housing.
![Screenshot (3)](https://user-images.githubusercontent.com/44751666/103696532-ceed0900-4f9e-11eb-8087-46ab3e6f8838.png)
![Screenshot (4)](https://user-images.githubusercontent.com/44751666/103696541-d1e7f980-4f9e-11eb-9e4f-a91791d2c9ca.png)
![Screenshot (5)](https://user-images.githubusercontent.com/44751666/103696555-d7ddda80-4f9e-11eb-854b-e852747a0198.png)

The pressure gauge will use a Honeywell i2c gage pressure sensor capable of up to 4 bar (60 PSI).
The temperature sensor will either be an NTC thermistor or a K-type thermocouple (yet to be decided).
Both gauges have a pause button to hold the reading.

All functions in the thermal camera settings work, all standard FLIR color maps are included, the min and max temperatures can be manually set, or the auto button can be pressed and the camera will auto-scale the min and max temperature. The autoscale range can also be adjusted. The sensor is an AMG8833 8x8 thermal array, and is interpolated to either 18x15 with a boxsize of 14x12, 36x36 with a boxsize of 7x5, or 63x60 with a box size of 4x3, depending on the resolution setting. The thermal camera is also used to take a reading of the asphalt temperature for the data log.

The datalog saves data to SD in .txt format, which will allow easy use for excell and so on in the future.
There are 6 "car" folders, and for each car up to 12 logs can be saved (theretically unlimited amount of logs can be saved). Each log contains the 12 temperatures taken from the tires, plus date, time, ambient temperature, and asphalt temperature. Averages for front, rear, left and right are calculated, along with the delta from min and max temps of each tire. Each tire also displays a color for inside, middle, outside corresponding to the temperature. The colors are taken from the thermal color palette array.

To do list:
-add a global settings popup with a button on the home page, which includes manual date/time setting, °c or °f, and screensaver timer adjustment.
-add rotary encoder as failsafe (touch screen seems to be very temperature sensitive).
-perhaps add some method of screen calibration that can be carried out during runtime.
-fix some minor bugs
-Find an appropriate push button pressure release valve to incorporate into the design (for pressure gauge) 
-finish cad design

I would like to also look into some kind of smart phone interface, so that the log data can be accessed via bluetooth. Maybe even stream the thermal camera to the smart phone ( its only an array of 64 floats, it could be interpolated locally on the target device)

I also attach a copy of the project, if anyone would like to take a look at it. I would be happy to hear peoples thoughts.

[TireToolFinal.zip](https://github.com/ImpulseAdventure/GUIslice/files/5772532/TireToolFinal.zip)

**Comment by ImpulseAdventure** [comment]
Hey @MickTheMechanic — this is really fantastic work! Thank you very much for sharing screenshots of your project and the background details. Gives people some great ideas.

If I were still doing autox (BS/BSP) and track days I would have definitely put your device to good use as it really addresses a very important part of the data logging. I’ll be interested to see future updates as you make progress. All the best with the project!

**Comment by MickTheMechanic** [comment]
Thanks 🙏 
I haven’t been programming for very long, and it has taken a lot of time and effort to get this far. 
I also wanted to ask, is there some way to take screenshots from the device or the builder? I thought I read something about this, but I haven’t found it again.

**Comment by MickTheMechanic** [comment]
Been working a bit more on hardware in the last weeks, things are looking good. 
I will spend some more time on getting the button interface working now, It really is an awesome thing that you guys created here, I wouldnt have gotten this far without GUISlice and the Builder! Can't thank you guys enough!

![IMG_0322](https://user-images.githubusercontent.com/44751666/106392300-779d5580-63f1-11eb-8b90-a0097a65f1bf.jpeg)
![IMG_0323](https://user-images.githubusercontent.com/44751666/106392302-79ffaf80-63f1-11eb-9261-af21accf528b.jpeg)
![IMG_0326](https://user-images.githubusercontent.com/44751666/106392305-7cfaa000-63f1-11eb-8679-eb9daada7f16.jpeg)
![IMG_0333](https://user-images.githubusercontent.com/44751666/106392307-808e2700-63f1-11eb-85f2-1c7b5da3ecdd.jpeg)
![IMG_0334](https://user-images.githubusercontent.com/44751666/106392316-871c9e80-63f1-11eb-9f10-4afb934a80bd.jpeg)

**Reply by ImpulseAdventure** [reply]
Very nice clean build. It is exciting to see your idea come together!

**Comment by MickTheMechanic** [comment]
_A small update._

I redid the wiring to reduce the combined length of the cables. I was having some wierd problems with the I2C BUS, from what I understand if the cables/tracks are too long this can increase the capacitance and create problems, so I'm hoping this will help.

I have ordered a new touch screen, this time a 3.95" screen that should be compatible with the tft_espi drivers that have been optimised for the ESP32.

The housing has been remodeled a bit for better placement of components and to allow for the bigger screen.

I have also added haptics.

![IMG_0460](https://user-images.githubusercontent.com/44751666/108770307-7081e700-755a-11eb-8da6-b0d171bac9c0.jpeg)
![IMG_0470](https://user-images.githubusercontent.com/44751666/108770324-74ae0480-755a-11eb-8a9e-e10d1f5efde7.jpeg)
![IMG_0469](https://user-images.githubusercontent.com/44751666/108770338-78418b80-755a-11eb-8836-3ac1a999682a.jpeg)
![IMG_0468](https://user-images.githubusercontent.com/44751666/108770355-7d063f80-755a-11eb-8db0-8e3a2bf370f1.jpeg)
![IMG_0467](https://user-images.githubusercontent.com/44751666/108770382-82fc2080-755a-11eb-9516-d175bebb4797.jpeg)
![IMG_0479](https://user-images.githubusercontent.com/44751666/108770396-87283e00-755a-11eb-91df-e9b5dd762e32.jpg)



Some info about the Haptics. I am using a drv2605l haptics driver with the adafruit lib.

To trigger the haptics during touchscreen touch and release I use GSLC_TOUCH_DOWN_IN and GSLC_TOUCH_UP_IN. This does the job nicely. 

```

bool CbBtnCommon(void* pvGui, void *pvElemRef, gslc_teTouch eTouch, int16_t nX, int16_t nY)
{
  // Typecast the parameters to match the GUI and element types
  gslc_tsGui*     pGui     = (gslc_tsGui*)(pvGui);
  gslc_tsElemRef* pElemRef = (gslc_tsElemRef*)(pvElemRef);
  gslc_tsElem*    pElem    = gslc_GetElemFromRef(pGui, pElemRef);
  ShowFocus(&m_gui, false);
  if ( eTouch == GSLC_TOUCH_DOWN_IN ) {
    triggerPressHaptic();
  }
  if ( eTouch == GSLC_TOUCH_UP_IN ) {
    screensaverTimer = currentMillis;
    triggerReleaseHaptic();

    // From the element's ID we can determine which button was pressed.
    switch (pElem->nId) {


// .... //


 //<Button Enums !End!>
      default:
        break;

    }
  }
  return true;
}


```





I also added a few lines to the slider code to get some haptic feedback when using the slider widgets:

```

// Callback function for when a slider's position has been updated
bool CbSlidePos(void* pvGui, void* pvElemRef, int16_t nPos)
{
  gslc_tsGui*     pGui                    = (gslc_tsGui*)(pvGui);
  gslc_tsElemRef* pElemRef         = (gslc_tsElemRef*)(pvElemRef);
  gslc_tsElem*    pElem                = gslc_GetElemFromRef(pGui, pElemRef);
  int16_t         nVal;
  int16_t         xVal                       = nVal;                      // save value to compare later
  gslc_tsElemRef* pElemRefTmp = NULL;
  
  // From the element's ID we can determine which slider was updated.
  switch (pElem->nId) {
    //<Slider Enums !Start!>

    case COLOR_LIST_SCRL:
      // Fetch the slider position
      nVal = gslc_ElemXSliderGetPos(pGui, colorListScrl);
      // Update the textbox scroll position
      pElemRefTmp = gslc_PageFindElemById(pGui, THERMAL_SETTINGS, COLOR_LIST);
      gslc_ElemXListboxSetScrollPos(pGui, pElemRefTmp, nVal);
      break;
    //<Slider Enums !End!>
    default:
      break;
  }
  if (nVal != xVal){                                   //If slider value has changed trigger haptics
    triggerScrollHaptic(); 
  }
  return true;
}

```



Unfortunately I am having some trouble with the push buttons. I am using the following code, but its not working how it should. Any tips on how I could get this working better?




```

bool CbPinPoll(void* pvGui, int16_t* pnPinInd, int16_t* pnPinVal)
{

  // Sample all pin inputs
  btn_prev.read();
  btn_sel.read();
  btn_next.read();

  // Determine if any pin edge events occur
  // Trigger press haptics if button was depressed
  if      (btn_prev.wasPressed()) {
    triggerPressHaptic();
    *pnPinInd = PIN_PREV;
    *pnPinVal = 1;
  }
  else if (btn_sel.wasPressed())  {
    triggerPressHaptic();
    *pnPinInd = PIN_SEL;
    *pnPinVal = 1;
  }
  else if (btn_next.wasPressed()) {
    triggerPressHaptic();
    *pnPinInd = PIN_NEXT;
    *pnPinVal = 1;
  }
  else return false; // No pin event detected

  // If we reach here, then an pin event was detected
 //Read buttons again to see if button was released
  btn_prev.read();
  btn_sel.read();
  btn_next.read();

  if (btn_next.wasReleased()|btn_sel.wasReleased()|btn_prev.wasReleased()) { // if yes, trigger release haptics
    triggerReleaseHaptic();
  }
  
  return true;

}

```

The goal would be to trigger a feedback response twice, once if the button is depressed, and another when the button is released.

**Reply by ImpulseAdventure** [reply]
Looking great!!

The haptics driver is a neat addition for UI feedback -- glad to see it is relatively easy to interface.

As for your PinPoll code, I think the issue is that you won't reach the ReleaseHaptic() clause since the pin state won't have changed between the first read() and the second read(). So, we have three scenarios to consider: a) button first pressed, b) button continued to be held down and c) button released. You want to call triggerPressHaptic() on (a) and triggerReleaseHaptic() on (c). However, the above code will probably exit at the return false clause on (c) without checking for the wasReleased() state.

Instead, I'm wondering if something like this might work better?
```c++
bool CbPinPoll(void* pvGui, int16_t* pnPinInd, int16_t* pnPinVal)
{

  // Sample all pin inputs
  btn_prev.read();
  btn_sel.read();
  btn_next.read();

  // Determine if any pin edge events occur
  // Trigger press haptics if button was depressed
  if      (btn_prev.wasPressed()) {
    triggerPressHaptic();
    *pnPinInd = PIN_PREV;
    *pnPinVal = 1;
    return true; // Notify GUI of action
  }
  else if (btn_sel.wasPressed())  {
    triggerPressHaptic();
    *pnPinInd = PIN_SEL;
    *pnPinVal = 1;
    return true; // Notify GUI of action
  }
  else if (btn_next.wasPressed()) {
    triggerPressHaptic();
    *pnPinInd = PIN_NEXT;
    *pnPinVal = 1;
    return true; // Notify GUI of action
  }
  else if (btn_next.wasReleased() || btn_sel.wasReleased() || btn_prev.wasReleased()) {
    // // Trigger release haptics
    triggerReleaseHaptic();
    return false; // No GUI action to take
  }

}

```

**Reply by MickTheMechanic** [reply]
Thanks for the reply! 
Since my last comment I have recieved the new display and I have everything working with the tft_espi library now. 
I can say performance is much better. Storing images in spiffs results in lightning fast load times, the thermal camera has picked up a few FPS, and its much nicer to have some extra space.


I have begun rebuilding the GUI to suit the bigger display.

![Image (7)](https://user-images.githubusercontent.com/44751666/109422976-0b0c7b00-79de-11eb-9857-0ada072d300d.jpeg)
![Image (6)](https://user-images.githubusercontent.com/44751666/109422977-0e076b80-79de-11eb-8bb4-025efeeb142b.jpeg)







I have a question regarding the base page.

Is there an easy way to omit the base page completly from a particular page?

I would like to show the base page widgets on every page _except for_ E_PG_MAIN. There are quite a few, so im wondering if its easier to omit the base page completely from the main page rather than changing the visability for each individual widget.

**Reply by ImpulseAdventure** [reply]
Hi @MickTheMechanic -- that's great news on the updated hardware!

Yes, if you want to hide the base page widgets, I believe you should be able to do it with the following:
```c++
// When current page is on home page, hide & disable "base" layer
gslc_SetStackState(pGui, GSLC_STACK_BASE, false, false);
...
// When current page is not on home page, show & enable "base" layer
gslc_SetStackState(pGui, GSLC_STACK_BASE, true, true);
```

**Reply by MickTheMechanic** [reply]
Thanks! I will test it out.

I have a question about text fill. I noticed since working with the target device set to ESP32/ using the tft_espi drivers I am having some issues with it. I am not sure if you are aware, or if there is some kind of setting I missed. Seems to affect most, if not all fonts. In my example I have some text that I added in the builder (5 characters), and I have set up the external storage space to hold 7 characters. The width of the fill seems right, but the height is correct for the width of the 5 characters. Am I missing something, or should I open an issue?

![Image (8)](https://user-images.githubusercontent.com/44751666/109432973-1bd4e500-7a0e-11eb-9649-cc583b454d18.jpeg)
[3.95V2.zip](https://github.com/ImpulseAdventure/GUIslice/files/6057928/3.95V2.zip)

I am also curious, is there any straightforward way that you're aware of to take a screen shot from the display?

Thanks again!

**Reply by ImpulseAdventure** [reply]
What your image appears to show is that the text field dimension estimated by the Builder doesn't quite match the actual region filled by TFT_eSPI's font rendering routines. It looks like TFT_eSPI might be leaving some room for descenders below the baseline. If you could try changing one of your numeric fields to include a letter like 'y', we may be able to confirm.

As for screenshots, yes, it is possible to capture the TFT image, but only on certain "readable" displays.

If you could attach your TFT_eSPI User_Setup file, it may help narrow down the font modes and also indicate what "screenshot" options may be available.

thanks

**Reply by MickTheMechanic** [reply]
Thanks for the reply, I have been working on a side project that was supposed to be "quick" but is taking a bit longer than expected. Unfortunately I havent come any further with this, but I will spend some more time on it during the week, and also test out your dev branch.

Here is my user setup file:
[custom_setup.zip](https://github.com/ImpulseAdventure/GUIslice/files/6103652/custom_setup.zip)

This is the display I am currently using:
http://www.lcdwiki.com/3.95inch_Arduino_Display-UNO?spm=a2g0o.detail.1000023.16.41845c50is0F3w&file=3.95inch_Arduino_Display-UNO

I appreciate the help.
